### ISSUE TITLE
Locking of build dependencies

### ISSUE THREAD

**Issue by sbidoul** [astral-sh/uv#5190]
Coming from https://github.com/astral-sh/uv/issues/3611#issuecomment-2226921962 and following comments.

Caveat: this is certainly not a complete proposal, and I reckon I have not investigated the capabilities of uv lock/sync in any detail, so consider this as thoughts to start the conversation on the usefulness of such features.

**What**

The `uv.lock` format supports locking source distributions (from sdists, git source trees, ...).

I would find it useful that
- commands that install from a lock file fail (by default?) if they need to install build dependencies that are not locked
- the lock command grows a mean to lock the build system requirements of the source dependencies declared in the lock file it produces

**Why**

When installing from a lock file I expect that everything that the installation downloads is checked against the hashes (or immutable VCS commit hashes) that are present in the lock file, for repeatability and security.

If, during installation, other artifacts are downloaded (or obtained from cache) without my knowledge (to build source distributions), the security and repeatability of the process is weakened. 

**Some thoughts**

- this may have performance implications for the lock command if sdists have to be downloaded to know the build requirements (https://github.com/astral-sh/uv/issues/3611#issuecomment-2226993337)
- some build dependencies are dynamic (`get_requires_for_build_{wheel,editable}`), further complicating the locking operation
- a first valuable step could be to lock build dependencies for which no wheel is available, and warn when an artifact that is not locked is downloaded when syncing

**Comment by paveldikov** [comment]
I was going to raise this same issue as well. Currently `build-system.requires` is completely unconstrained/unconstrainable and this is causing problems with build reproducibility.

This is causing real-life inconveniences in our organisation as, for example, new versions of `setuptools` appear but are not yet given the all-clear by our scanning suite. Versions that are not cleared yield a HTTP 403 upon download. The end result is a de-facto intractable state where no one can build (or even editably install) packages. (unless the developer decides to hard-constrain @ `pyproject.toml`, which is terrible UX and also doesn't help with the dependency-on-sdist problem.)

(I will probably also raise a separate issue w.r.t the handling of 403 codes -- this is a pain point with `pip` and it would be awesome if `uv` could do better)

**Comment by charliermarsh** [comment]
Definitely agree that we should do this.

**Comment by charliermarsh** [comment]
I really want to do this (lock the build requirements).

**Comment by charliermarsh** [comment]
I really want to do this and would like to spend time on it personally, but I'm worried about how it will intersect with PEP 751.

**Comment by notatallshaw** [comment]
> I really want to do this and would like to spend time on it personally, but I'm worried about how it will intersect with PEP 751.

Last I saw build time dependencies were dropped from PEP 751: https://discuss.python.org/t/pep-751-lock-files-again/59173/222 (end of that post).

Did something change?

**Comment by charliermarsh** [comment]
@notatallshaw -- No, I agree with you, it's just that: if we add support for it here, what happens if PEP 751 comes out and we can't translate that support to the PEP?

**Comment by bluss** [comment]
Could you use the space for tool specific data in the PEP 751 lock file? Or even using one of the main fields like `groups` using some kind of convention that makes it possible for locked build deps to co-exist with other named things stored there.

**Comment by notatallshaw** [comment]
Yeah, makes sense to see the finalized version of PEP 751 first.

**Comment by dimbleby** [comment]
similar proposals in poetry, pdm: https://github.com/python-poetry/poetry/issues/8261 (and several duplicates), https://github.com/pdm-project/pdm/issues/2465

I mostly think that people who care about reproducible builds would be better served by avoiding sdists altogether with `--only-binary` or similar.

**Comment by giftig** [comment]
+1 for this; I've just had an existing, locked dependency fail to install because a new version of `setuptools` refuses to install dependencies using deprecated `description-file` syntax in `setup.cfg`, which is confusing and slightly frustrating (though not the fault of `uv`)

**Comment by sbidoul** [comment]
Yep, to me this would be the killer feature to switch to `uv.lock`.

Build contraints help but they'd need to be managed manually, won't work if there are conflicting build constraints in the project, and it's easy to forget a build constraint, which you only realize months years later when a build fails.

**Comment by purepani** [comment]
Since PEP 751 is now accepted, and the resulting lock file isn't immediately usable for `uv` as a project lock file, perhaps it makes sense to add this as a `uv.lock` feature.

**Comment by charliermarsh** [comment]
We're hoping to start working on this soon.

**Comment by jiridanek** [comment]
> Build contraints help but they'd need to be managed manually, won't work if there are conflicting build constraints in the project, and it's easy to forget a build constraint, which you only realize months or years later when a build fails.

Building with `--require-hashes` seems to be a decent check for a forgotten build constraint. With

```
❯ uv --version
uv 0.8.5
```

I get, when I use `--build-constraints=./requirements.txt --no-binary=:all:`

```
❯ uv pip install --strict --no-deps --no-cache --no-config --no-progress --require-hashes --index-strategy=unsafe-best-match --requirements=./requirements.txt --build-constraints=./requirements.txt --no-binary=:all:
Using Python 3.12.11 environment at: pepa
Resolved 71 packages in 268ms
   Building nbclient==0.10.2
   Building ptyprocess==0.7.0
   Building typing-extensions==4.14.1
   Building jupyter-core==5.7.2
   Building tinycss2==1.4.0
  × Failed to download and build `nbclient==0.10.2`
  ├─▶ Failed to resolve requirements from `build-system.requires`
  ├─▶ No solution found when resolving: `hatchling>=1.10.0`
  ╰─▶ In `--require-hashes` mode, all requirements must be pinned upfront with `==`, but found: `hatchling`
```

The above is actually a great inconvenience to me, but I see some may benefit from this behavior.

**Comment by bruno-fs** [comment]
I find it a bit sad that new "universal lockfile format" from PEP 751 [won't support locking build requirements for sdists](https://peps.python.org/pep-0751/#locking-build-requirements-for-sdists), but I'm glad to hear uv is interested in solving this problem.

**Comment by tleonhardt** [comment]
My organization needs this capability to support offline build environments.

**Comment by jiridanek** [comment]
I'm having a Chief Architect asking me questions because we could not pin our transitive build-dependencies and meson 1.11 release yesterday broke the pandas 2.3 build
* https://github.com/pandas-dev/pandas/pull/63406
